### ISSUE TITLE
Che workspace is stuck in stopping state and cannot be deleted (prod-prev)

### ISSUE THREAD

**Issue by ljelinkova** [openshiftio/openshift.io#4391]
##### Issue Overview
On our prod preview account for E2E tests (osio-ci-e2e-001-preview) we have one workspace that is stuck in stopping state. It is not possible to delete it and it is preventing any other workspace to start. 

We tried `oc delete all` as suggested in https://github.com/openshiftio/openshift.io/issues/4379 but it did not help. 

Could it be that the user does not have access rights to delete it via Openshift? Could anybody delete the workspace for us?

**Comment by ppitonak** [comment]
@l0rd @ibuziuk @rhopp @Katka92  could you have a look?

**Comment by sleshchenko** [comment]
Unfortunately, removing resources on OpenShift cluster is not enough. One more thing that should be cleaned up too is Runtime entity in Che Server Database.
I'm working on an improvement to make Che Server cleaning up such workspace by itself https://github.com/eclipse/che/issues/11364#issuecomment-425383803

**Comment by ppitonak** [comment]
It looks like we know the root cause, can somebody manually fix our account?

**Comment by l0rd** [comment]
@ppitonak That's the first time I hear that. I mean that `oc delete all` does not remove the resource. @sleshchenko with `oc delete` the workspace pod should be removed and wsmaster won't recreate it right? Maybe [che namespace lockdown](https://gitlab.cee.redhat.com/dtsd/housekeeping/issues/844) has been enforced on that account.

**Comment by sleshchenko** [comment]
> @sleshchenko with oc delete the workspace pod should be removed and wsmaster won't recreate it right?

Right. But even after OpenShift resources removing (pods, services, routes) Che Server still consider workspace as STOPPING and doesn't allow to remove it.

**Comment by garagatyi** [comment]
Looks like the only way to fix the account at the moment is to delete it from DB manually. @sleshchenko WDYT?

**Comment by ljelinkova** [comment]
We have another account with workspace stuck in stopping state on production (osio-ci-e2e-001). It happend 11 hours ago.

**Comment by sleshchenko** [comment]
> Looks like the only way to fix the account at the moment is to delete it from DB manually. @sleshchenko WDYT?

I think removing the corresponding entities from DB manually is the only way. I can provide needed SQL queries.

**Comment by sleshchenko** [comment]
> We have another account with workspace stuck in stopping state on production (osio-ci-e2e-001). It happend 11 hours ago.

@garagatyi @amisevsk
It makes sense to investigate why it happened. If Che Server crashed in that time then it will be fixed by my next PR. Otherwise, the separated issue should be created.
Do you have access to Che Server Logs?

**Comment by ljelinkova** [comment]
I do not have access but @rhopp was looking into it. The 'stuck' workspace name is yims9. 

I am not sure if this is related but I noticed that there were many workspaces from previous runs of E2E tests (~60). We do reset environment after every test so I'd expect there to be 0 workspaces after reset.

**Comment by ljelinkova** [comment]
Have you done anything with the prod account? Because it seems to be fixed now. The tests started to pass at 4pm UTC. 

The prod-preview account is still stuck though.

**Comment by garagatyi** [comment]
No, we haven't fixed it. We don't know the reason and still investigating how to fix intermittent occurrence.

**Comment by ldimaggi** [comment]
Regarding the severity of this issue -if the problem is 100% solved in production and the E2E tests are passing, then we can move it to SEV2. The P1 designation should remain if the issue as presented on prod-preview is causing the E2E tests to fail (on prod-preview).

**Comment by ldimaggi** [comment]
The issue of multiple workspaces remaining after an attempt to delete them is a different issue. I saw that same problem earlier this week. The symptoms of that issue were that a workspace was "stuck" which seems to cause other workspaces to not be deleted. I have not seen this issue a 2nd time (yet).

**Comment by garagatyi** [comment]
There is a [fix](https://github.com/eclipse/che/pull/11570) in upstream that should fix the situation. It removes the state of a hanging workspace after ~8 minutes of waiting for the workspace to report its successful stopping. Hopefully, we would be able to deliver this fix with a bugfix release of upstream soon.
It would be great to understand the reason why workspace got this state to fix the root cause. It is probably related to an OOM of Che server or pod restart because of an update. 
Can somebody provide workspace IDs for those workspaces that got this state on prod and prod-preview?

**Comment by garagatyi** [comment]
Hopefully days only.

**Comment by ibuziuk** [comment]
@ljelinkova could you please tell if the prod-preview account is now fixed ? the fix [1] has been available on prod-preview / prod for a while.

[1] https://github.com/eclipse/che/pull/11570

**Comment by ljelinkova** [comment]
No, unfortunately, the workspace is still `stopping`.

**Comment by rhopp** [comment]
I don't think that fix is already in prod/prod-preview. The PR was merged 7 days ago, 6.12.2. was released 8 days ago... Furthermore I don't see the commits from that PR in commits Alex cherrypicked to 6.12.2 release. I think it will be available when we switch to 6.13.

**Comment by ibuziuk** [comment]
@rhopp @ppitonak could you please point me to prod-preview e2e ci job which is failing due to this issue? AFAIK, prod e2e tests are not affected by this issue:
- https://ci.centos.org/job/devtools-saas-openshiftio-e2e-smoketest-released/
- https://ci.centos.org/job/devtools-saas-openshiftio-e2e-smoketest-beta/

**Comment by ppitonak** [comment]
This is the job https://ci.centos.org/job/devtools-test-e2e-prod-preview.openshift.io-smoketest-pr-us-east-2a-released/

This is how it looks like for end user
![che](http://artifacts.ci.centos.org/devtools/e2e/devtools-test-e2e-prod-preview.openshift.io-smoketest-pr-us-east-2a-released/3245/04-01-afterEach.png)

**Comment by sleshchenko** [comment]
@rhopp 
> I don't think that fix is already in prod/prod-preview. The PR was merged 7 days ago, 6.12.2. was released 8 days ago... Furthermore I don't see the commits from that PR in commits Alex cherrypicked to 6.12.2 release. I think it will be available when we switch to 6.13.

Thanks for a good news for me. Because I was just wondering why my fix does solve this issue.
So, let me know if it doesn't solve issue after updating to 6.13

**Comment by ibuziuk** [comment]
@ljelinkova @ppitonak @rhopp 6.13.0 has been released on prod-preview could you please check if this problematic workspace is still in the stopping phase and can not be deleted ?

**Comment by ljelinkova** [comment]
The fix worked and the workspace is now stopped. 

![workspace-eigje](https://user-images.githubusercontent.com/2078072/47498666-3c993880-d85e-11e8-8c51-bd56e1966dde.png)

Let's wait until the next e2e test (scheduled in 10 minutes) and if it goes well I'll close this issue.
